internal/lock/lock.go

```diff
@@ -0,0 +1,189 @@
+package lock
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+
+	"go.etcd.io/etcd/clientv3"
+)
+
+const (
+	keyTemplate  = "com.coreos.airlock/groups/%s/v1/semaphore"
+	defaultGroup = "default"
+)
+
+var (
+	// ErrNilManager is returned on nil manager
+	ErrNilManager = errors.New("nil Manager")
+)
+
+// Manager takes care of locking for clients
+type Manager struct {
+	client  *clientv3.Client
+	keyPath string
+}
+
+// NewManager returns a new lock manager, ensuring the underlying semaphore is initialized.
+func NewManager(ctx context.Context, etcdURLs []string, group string, slots uint64) (*Manager, error) {
+	// TODO(lucab): move to clientv3.New(clientv3.Config)
+	client, err := clientv3.NewFromURL(etcdURLs[0])
+	if err != nil {
+		return nil, err
+	}
+
+	keyPath := fmt.Sprintf(keyTemplate, url.QueryEscape(group))
+	manager := Manager{client, keyPath}
+
+	if err := manager.ensureInit(ctx, slots); err != nil {
+		return nil, err
+	}
+
+	return &manager, nil
+}
+
+// RecursiveLock adds this lock `id` as a holder of the semaphore
+//
+// It will return an error if there is a problem getting or setting the
+// semaphore, or if the maximum number of holders has been reached.
+func (m *Manager) RecursiveLock(ctx context.Context, id string) error {
+	sem, version, err := m.get(ctx)
+	if err != nil {
+		return err
+	}
+
+	held, err := sem.RecursiveLock(id)
+	if err != nil {
+		return err
+	}
+	if held {
+		return nil
+	}
+
+	if err := m.set(ctx, sem, version); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// UnlockIfHeld removes this lock `id` as a holder of the semaphore
+//
+// It returns an error if there is a problem getting or setting the semaphore.
+func (m *Manager) UnlockIfHeld(ctx context.Context, id string) error {
+	sem, version, err := m.get(ctx)
+	if err != nil {
+		return err
+	}
+
+	if err := sem.UnlockIfHeld(id); err != nil {
+		return err
+	}
+
+	if err := m.set(ctx, sem, version); err != nil {
```

Comment:
Would it be worthwhile to add detection (maybe in a future PR) if the semaphore is already in the desired state and return early rather than potentially running a no-op `set`?

Reply:
Indeed, it seems sound and better performance-wise in order to reduce latency and etcd contention. Tracked at https://github.com/coreos/airlock/issues/9.